text/0000-lifecycle-experimental-api-mode.md

```diff
@@ -0,0 +1,122 @@
+[apis.buildpack]
+  experimental = ["0.6-rc1", "1.0-alpha1"]
+[apis.platform]
+  experimental = ["1.0-alpha1"]
+
+[lifecycle]
+  version = "0.10.0"
+```
+`experimental`:
+contains an array of experimental API versions
+the APIs will correspond to specs that may not be released and in branches that are still being worked on.
+the API version must be of the format: `<major>.<minor>-<alphanumeric>`, i.e. `1.0-alpha1`.
+
+Since experimental versions are defined with a different format, they will NOT automatically get upgraded to a `supported` release when the experimental API has finalized.
+
+## Lifecycle Labels
+This will extend the builder or lifecycle image with the above descriptor file will extend the label `io.buildpacks.lifecycle.apis` with the value:
+```json
+{
+  "buildpack": {
+    "experimental": ["0.6-rc1", "1.0-alpha1"],
+  },
+  "platform": {
+    "experimental": ["1.0-alpha1"],
+  }
+}
+```
+
+## API Types
+### Experimental APIs
+API versions aren't limited to those defined in a spec release, but can refer to upcoming spec work happening in a branch.
+
+New `CNB_PLATFORM_EXPERIMENTAL_MODE` and `CNB_BUILDPACK_EXPERIMENTAL` mode environment variables will control experimental mode with:
+allowed values: `warn`, `error`, `silent`
+default value: `warn`
+
+**When** the `CNB_PLATFROM_API` environment variable is set to an API version in the deprecated platform API, the lifecycle shall:
+ - **If** `CNB_PLATFORM_EXPERIMENTAL_MODE` is unset, **Then** print a warning and continue
+ - **If** `CNB_PLATFORM_EXPERIMENTAL_MODE=warn`, **Then** print a warning and continue
+ - **If** `CNB_PLATFORM_EXPERIMENTAL_MODE=error`, **Then** fail
+ - **If** `CNB_PLATFORM_EXPERIMENTAL_MODE=silent`, **Then** continue w/o warning
+
+**When** the `api` field in a `buildpack.toml` file is set to an API version in the deprecated buildpack API range the lifecycle shall:
+ - **If** `CNB_BUILDPACK_EXPERIMENTAL_MODE` is unset, **Then** print a warning and continue
+ - **If** `CNB_BUILDPACK_EXPERIMENTAL_MODE=warn`, **Then** print a warning and continue
+ - **If** `CNB_BUILDPACK_EXPERIMENTAL_MODE=error`, **Then** fail
+ - **If** `CNB_BUILDPACK_EXPERIMENTAL_MODE=silent`, **Then** continue w/o warning
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The technical implementation will follow what's outlined in [RFC#49](https://github.com/buildpacks/rfcs/blob/main/text/0049-multi-api-lifecycle-descriptor.md).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+This will add more complexity/overhead that most users will not use.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Do Nothing
+If we choose to do nothing, we run the risk of releasing less tested API/spec releases.
+
+### Not a Different Version Format
+The Experimental API version format could instead follow the same format as deprecated and supported: `<major>-<minor>`. By doing this, users could automatically get upgraded from an experimental to supported/deprecated API version which may be surprising to them.
```

Comment:
@ekcasey is your question about this alternative, or did the alternative prompt you to ask this about the main proposal.

Where would we be removing the experimental version from?

Reply:
> @ekcasey is your question about this alternative, or did the alternative prompt you to ask this about the main proposal?

The latter.

> Where would we be removing the experimental version from?

The `experimental` section of the next lifecycle release. I assume we wouldn't want to support every experimental API forever especially if we finalize and release the feature in a supported API?

Reply:
So you're saying when `lifecycle` releases 1.0, we would remove Platform/Buildpack API support for 1.0-alpha1?

I guess I was thinking it would be more like when `lifecycle` releases 1.0, we would remove Platform/Buildpack API support for 0.9-alpha1. That is:
* for any `<major>.<minor>` we would also support `<major>.<minor>-<experimental>`. 
* when we drop a given `<major>.<minor>` we would also drop the corresponding `<major>.<minor>-<experimental>`

Reply:
I agree with your description of what would happen, but we still remove the experimental version eventually right? Do we deprecate the experimental version before we remove it?

I guess this drawback confuses me but it implies some things I am necessarily comfortable with.

In the original multi-api RFC the idea of "experimental support" was that the definition of a experimental API was subject to change. I think this is now implying that the definition of `1.0-alpha1` is never allow to change, instead we would need to add a `1.0-alpha2`? If that is true
1. Do we really want to do that? The lifecycle could end up supporting many more APIs concurrently, adding complexity
2. Do "experimental" APIs even make sense as a category of APIs? Wouldn't it make more sense then to say that `1.0-alpha1` is first "supported" and later "deprecated" in that case?

Reply:
> Do "experimental" APIs even make sense as a category of APIs? Wouldn't it make more sense then to say that 1.0-alpha1 is first "supported" and later "deprecated" in that case?

Like I said to @jromero's questions above, I'm fine with experimental not being its own section, but I do believe unlike both of those it carries a different support guarantees/breaking policies. This just makes that clearer by making it separate.

> I think this is now implying that the definition of 1.0-alpha1 is never allow to change, instead we would need to add a 1.0-alpha2

My hope is that we could branch/tag a spec, but maybe that overhead isn't worth it.

> Do we really want to do that? The lifecycle could end up supporting many more APIs concurrently, adding complexity

The key is to have a good/fast way to sunset experimental modes IMO.